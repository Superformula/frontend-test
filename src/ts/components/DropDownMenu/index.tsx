// Imports
// -------

// Libraries
import React from 'react';
// Components
import Checkbox from '../Checkbox';
// Images
import carat from '../../../img/carat.svg';
// Style
import style from './style.module.css';

// Internal
// --------

export interface DropDownMenuProps {
    text?: string,
    children: string[],
    selected?: boolean,
    dimensions?: {
        width: number,
        height: number
    },
    option?: string | null,
    setOption?(arg0: string): any,
    onClick?(): any
}

const DropDownMenu:React.FC<DropDownMenuProps> = ({text='', children, selected = false, dimensions = {}, onClick = () => {},  option = null, setOption = (arg0: string) => {}}) => {
    //
    // Render

    let shownText = text;
    if (option) {
        shownText = option;
    }

    const getListItems = (child: string, idx: number) => (
            <li key={idx}>
                <Checkbox checked={ option === child } onClick={ () => setOption(child) }>
                    {child}
                </Checkbox>
            </li>
    );

    const listClassName = [style.dropDownMenuList, selected ? style.show : ''].join(' ');
    const menuStyle = {
        width: dimensions.width,
        height: dimensions.height
    };

    return (
        <div className={ style.dropDownMenu } onClick={ onClick } style={ menuStyle }>
            <div className={ style.selectedText }>
                <span>{ shownText }</span>
                <img src={ carat } alt='dropdown button' className={ [style.carat, selected ? style.rotate : ''].join(' ') }/>
            </div>
            <ul className={ listClassName }>
                { React.Children.map(children, getListItems) }
            </ul>
        </div>
    );
}

// Exports
// -------

export default DropDownMenu;
