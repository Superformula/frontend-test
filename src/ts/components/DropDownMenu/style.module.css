.dropDownMenu {
    position: relative;
    user-select: none;
}

.selectedText {
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    padding: 8px 0;
}

.dropDownMenuList {
    background: #fff;
    border: 1px solid #C8C8C8;
    border-top-width: 0;
    box-shadow: none;
    box-sizing: border-box;
    list-style: none;
    margin: 0;
    max-height: 0;
    overflow: hidden;
    padding: 0 16px;
    position: absolute;
    transition-duration: .4s;
    transition-property: all;
    transition-timing-function: ease-out;
    width: 100%;
    z-index: 2;
}

.show {
    border: 1px solid #C8C8C8;
    box-shadow: 0 6px 5px rgba(0,0,0,0.1), inset 0 -1px 0 #C8C8C8;
    max-height: 300px;
}

.carat {
    transition-duration: 0.5s;
    transition-timing-function: ease-out;
}

.rotate {
    transform: rotate(-180deg);
}
