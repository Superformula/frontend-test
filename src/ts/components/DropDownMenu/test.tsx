// Imports
// -------

// Libraries
import React from 'react';
import renderer from 'react-test-renderer';
// Components
import DropDownMenu from './index';

// Mocks
// -----

jest.mock("../Checkbox", () => 'Checkbox');
jest.mock('../../../img/carat.svg', () => 'carat.svg');

// Internal
// --------

const children = ['Opt1', 'Opt2', 'Opt3'];

const getComponent = (props = {}) => renderer.create(
    <DropDownMenu {...props}>
        { children }
    </DropDownMenu>,
);

// Tests
// -----

describe('src/ts/components/DropDownMenu', () => {
    let component;

    describe('Rendering', () => {
        it('renders default correctly', () => {
            component = getComponent();
            expect(component).toMatchSnapshot();
        });
    });
});
