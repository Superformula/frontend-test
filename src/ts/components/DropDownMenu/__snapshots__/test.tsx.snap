// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/ts/components/DropDownMenu Rendering renders default correctly 1`] = `
<div
  className="dropDownMenu"
  onClick={[Function]}
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <div
    className="selectedText"
  >
    <span>
      
    </span>
    <img
      alt="dropdown button"
      className="carat "
      src="carat.svg"
    />
  </div>
  <ul
    className="dropDownMenuList "
  >
    <li>
      <Checkbox
        checked={false}
        onClick={[Function]}
      >
        Opt1
      </Checkbox>
    </li>
    <li>
      <Checkbox
        checked={false}
        onClick={[Function]}
      >
        Opt2
      </Checkbox>
    </li>
    <li>
      <Checkbox
        checked={false}
        onClick={[Function]}
      >
        Opt3
      </Checkbox>
    </li>
  </ul>
</div>
`;
